### PR TITLE
fix Issue 24155 - ImportC: accept C23 default initializers

### PR DESCRIPTION
**Pull request by WalterBright**
Did it by adding a `DefaultInitializer` AST node. Mostly boilerplate.

### CODE REVIEW

**Comment by ibuclaw** [comment]
Is there a comprehensive action list for supporting C23? There's a _lot_ of new additions and we shouldn't be doing it piecemeal because someone raised an issue about only one of dozens of changes necessary.

File: compiler/src/dmd/cparse.d [hunk 1]
```diff
@@ -2168,6 +2168,7 @@ final class CParser(AST) : Parser!AST
      * C11 Initialization
      * initializer:
      *    assignment-expression
+     *    { }                       // C23 6.7.10 addition
```
Comment (ibuclaw) [inline compiler/src/dmd/cparse.d:2171]: I understood C23 syntax to be `int var = ();` - this needs to be checked again.

**Review by ibuclaw — DISMISSED** [review]

File: compiler/src/dmd/cparse.d (hunk 1 above)
Comment (WalterBright) [reply compiler/src/dmd/cparse.d:2171]: It's `{ }`. See C23 6.7.10

**Comment by WalterBright** [comment]
@ibuclaw there isn't a comprehensive list. Support will be piecemeal regardless, based on the notion of things that are blocking people should come first, like all the other extensions. Apparently default initializers are commonly used.

**Comment by WalterBright** [comment]
@ibuclaw see https://en.wikipedia.org/wiki/C23_(C_standard_revision)#Syntax for default initializers

**Comment by ibuclaw** [comment]
> @ibuclaw there isn't a comprehensive list. Support will be piecemeal regardless, based on the notion of things that are blocking people should come first, like all the other extensions. Apparently default initializers are commonly used.

https://gcc.gnu.org/wiki/cauldron2023talks?action=AttachFile&do=view&target=ISO+C23+support+in+the+GNU+Toolchain.pdf

**Comment by WalterBright** [comment]
Thanks! It doesn't seem to be comprehensive - it's missing the `{ }` syntax. The Wikipedia entry I posted here seems better.

**Comment by ibuclaw** [comment]
> Thanks! It doesn't seem to be comprehensive - it's missing the `{ }` syntax. The Wikipedia entry I posted here seems better.

It just says `empty braces` -  I must have misunderstood whilst I was busy scribbling down everything.

There should be a video of the talk posted soon if not already.

**Comment by Imperatorn** [comment]
Thanks, it's common to use {} and {0}

**Comment by Imperatorn** [comment]
> Thanks, it's common to use {} and {0}

Both will work now right?

File: compiler/src/dmd/initsem.d [hunk 2]
```diff
@@ -1017,6 +1024,12 @@ Initializer inferType(Initializer init, Scope* sc)
         return new ErrorInitializer();
     }
 
+    Initializer visitDefault(DefaultInitializer i)
+    {
+        error(i.loc, "cannot infer type from default initializer");
```
Comment (maxhaton) [inline compiler/src/dmd/initsem.d:1029]:
```suggestion
        error(i.loc, "could not infer type from default initializer");
```
Or even better "A type could not be inferred from"

Comment (WalterBright) [reply compiler/src/dmd/initsem.d:1029]: I don't see an improvement?

**Comment by WalterBright** [comment]
`{0}` is not a generic default initializer and is not covered by this PR. I don't see it is a special case in C23, either.

**Comment by ibuclaw** [comment]
> > Thanks, it's common to use {} and {0}
> 
> Both will work now right?

`{}` and `{0}` don't mean the same thing if you are referring to initializing a struct.

**Comment by Imperatorn** [comment]
> > > Thanks, it's common to use {} and {0}
> > 
> > 
> > Both will work now right?
> 
> `{}` and `{0}` don't mean the same thing if you are referring to initializing a struct. It's already valid to have `int i = {0};`.

What I'm asking is, are both valid in ImportC with this change

**Comment by WalterBright** [comment]
`{0}` was already valid before this change. This changed added `{}`.

**Comment by WalterBright** [comment]
What is the reluctance to not approve this?

**Comment by Imperatorn** [comment]
Feel the groove and hit approve 🎊

**Review by dkorpel — APPROVED** [review]
I presume simply rewriting `{}` to `{0}` in the parser doesn't work?

**Comment by WalterBright** [comment]
Every time I try something like that I eventually wind up in a corner case somewhere.

**Pull request merged**